File: server-python/src/v1/blueprints/employee.py
from sanic.blueprints import Blueprint
from sanic.response import json, html
from ..database.employee import Qeury
from datetime import datetime as dt
from os.path import join, dirname, abspath

bp = Blueprint('employee', url_prefix='/pvd')
time_format = '%Y-%m-%d'
with open(join(dirname(abspath(__file__)), 'template.tphtml'), 'r') as f:
    template = f.read()
    
@bp.route('/', methods=["GET"])
async def get_all_employee(request):
    data = Qeury.get_all_employee()
    reponse = [{} for i in range(len(data))]
    now = dt.now()
    for i, d in enumerate(data):
        reponse[i]['Employee'] = d
        reponse[i]['TotalPVD'] = calculate_total_pvd(d, now)
        reponse[i]['TotalMonthPVD'] = calculate_total_month_pvd(d, now)
    return json(reponse)

@bp.route('/web', methods=["GET"])
async def get_all_employee(request):
    data = Qeury.get_all_employee()
    reponse = [{} for i in range(len(data))]
    now = dt.now()
    for i, d in enumerate(data):
        reponse[i]['Employee'] = d
        reponse[i]['TotalPVD'] = calculate_total_pvd(d, now)
        reponse[i]['TotalMonthPVD'] = calculate_total_month_pvd(d, now)
    data_str = ''
    for i, d in enumerate(reponse):
        data_str += '<tr>\n'
        data_str += f'<td>{d["Employee"]["ID"]}</td>\n'
        data_str += f'<td>{d["Employee"]["FirstName"]}</td>\n'
        data_str += f'<td>{d["Employee"]["LastName"]}</td>\n'
        data_str += f'<td>{d["Employee"]["BirthDate"]}</td>\n'
        data_str += f'<td>{d["Employee"]["StartDate"]}</td>\n'
        data_str += f'<td>{d["Employee"]["Salary"]}</td>\n'
        data_str += f'<td>{d["Employee"]["PvdFundRate"]}</td>\n'
        data_str += f'<td>{d["TotalPVD"]}</td>\n'
        data_str += '</tr>\n'
    return html(template.format(tbody=data_str))

def calculate_pvd(employee, month, paid_rate):
    return ((employee['Salary'] * paid_rate / 100) * month) + ((employee['Salary'] * employee['PvdFundRate'] / 100) * month)

def calculate_total_pvd(employee, time_now):
    total_pvd = 0
    t =  dt.strptime(employee['StartDate'], time_format)
    diff = time_now - t
    year = diff.days // 365
    month = diff.days % 365 // 30
    if year >= 5: # over 5 year
        total_pvd += calculate_pvd(employee, 12*(year-5)+month, 80)
        year = 5
        month = 0
    if year >= 3: # less than 5 year
        total_pvd += calculate_pvd(employee, 12*(year-3)+month, 50)
        year = 3
        month = 0
    if year >= 1: # less than 3 year
        total_pvd += calculate_pvd(employee, 12*(year-1)+month, 30)
        year = 1
        month = 0
    total_pvd += calculate_pvd(employee, max((12*year)+month-3, 0), 10) # less than 1 year
    return total_pvd

def calculate_total_month_pvd(employee, time_now):
    total_pvd = 0
    t =  dt.strptime(employee['StartDate'], time_format)
    diff = time_now - t
    year = diff.days // 365
    month = diff.days % 365 // 30
    total_month = (12*year + month) - 3 
    if total_month <= 0:
        return 0
    if year >= 5: # over 5 year
        return calculate_pvd(employee, total_month, 80)
    elif year > 3: # less than 5 year
        return calculate_pvd(employee, total_month, 50)
    elif year > 1: # less than 3 year
        return calculate_pvd(employee, total_month, 30)
    else: # less than 1 year
        return calculate_pvd(employee, total_month, 10)

File: server/controllers/employee.go
package controllers

import (
	"database/sql"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"time"

	"github.com/ballyees/mycos-assignment/database/model"
	"github.com/gofiber/fiber/v2"
	"github.com/valyala/fasttemplate"
)

const TimeFormat = "2006-01-02"

var tmpl = ReadFile("./controllers/template.html")
var rndr = fasttemplate.New(tmpl, "{{", "}}")

type EmployeePVD struct {
	Employee      model.Employee
	TotalPVD      float64
	TotalMonthPVD float64
}

func CalculatePVD(month float64, paidRate float64, employee model.Employee) float64 {
	return ((employee.Salary * paidRate / 100) * month) + ((employee.Salary * employee.PvdFundRate / 100) * month)
}

func CalculateTotalMonthPVD(employee model.Employee, now time.Time) (totalPVD float64) {
	t, _ := time.Parse(TimeFormat, employee.StartDate)
	diff := now.Sub(t)
	year := diff.Hours() / (24 * 365)
	month := math.Floor((year - math.Floor(year)) * 12)
	year = math.Floor(year)
	totalMonth := ((year * 12) + month) - 3
	if totalMonth <= 0 {
		return
	}
	if year > 5 { // over 5 year
		totalPVD = CalculatePVD(totalMonth, 80, employee)
	} else if year > 3 { // less than 5 year
		totalPVD = CalculatePVD(totalMonth, 50, employee)
	} else if year > 1 { // less than 3 year
		totalPVD = CalculatePVD(totalMonth, 30, employee)
	} else if year >= 0 { // less than 1 year
		totalPVD = CalculatePVD(totalMonth, 10, employee)
	}
	return
}

func CalculateTotalPVD(employee model.Employee, now time.Time) (totalPVD float64) {
	t, _ := time.Parse(TimeFormat, employee.StartDate)
	diff := now.Sub(t)
	year := diff.Hours() / (24 * 365)
	month := math.Floor((year - math.Floor(year)) * 12)
	year = math.Floor(year)
	if year >= 5 { // over 5 year
		totalPVD += CalculatePVD(12*(year-5)+month, 80, employee)
		year = 5
		month = 0
	}
	if year >= 3 { // less than 5 year
		totalPVD += CalculatePVD(12*(year-3)+month, 50, employee)
		year = 3
		month = 0
	}
	if year >= 1 { // less than 3 year
		totalPVD += CalculatePVD(12*(year-1)+month, 30, employee)
		year = 1
		month = 0
	}
	totalPVD += CalculatePVD(math.Max((12*year)+month-3, 0), 10, employee) // less than 1 year
	return totalPVD
}

func GetEmployeePVD(db *sql.DB) func(c *fiber.Ctx) error {
	return func(c *fiber.Ctx) error {
		err, data := QueryAllEmployee(db)
		if err != nil {
			return c.JSON(fiber.Map{
				"error": err.Error(),
			})
		}
		now := time.Now()
		employeePVDs := make([]EmployeePVD, len(data))
		for i, d := range data {
			employeePVDs[i].Employee = d
			employeePVDs[i].TotalPVD = CalculateTotalPVD(d, now)
			employeePVDs[i].TotalMonthPVD = CalculateTotalMonthPVD(d, now)
		}
		// time.Sleep(time.Duration(10) * time.Second)
		return c.JSON(employeePVDs)
	}
}

func GetEmployeePVDHtml(db *sql.DB) func(c *fiber.Ctx) error {
	return func(c *fiber.Ctx) error {
		err, data := QueryAllEmployee(db)
		if err != nil {
			return c.JSON(fiber.Map{
				"error": err.Error(),
			})
		}
		now := time.Now()
		employeePVDs := make([]EmployeePVD, len(data))
		for i, d := range data {
			employeePVDs[i].Employee = d
			employeePVDs[i].TotalPVD = CalculateTotalPVD(d, now)
			employeePVDs[i].TotalMonthPVD = CalculateTotalMonthPVD(d, now)
		}
		// time.Sleep(time.Duration(10) * time.Second)
		c.Set(fiber.HeaderContentType, fiber.MIMETextHTML)
		var reponse string

		for _, d := range employeePVDs {
			reponse += "<tr>\n"
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.ID)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.FirstName)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.LastName)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.BirthDate)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.StartDate)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.Salary)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.Employee.PvdFundRate)
			reponse += fmt.Sprintf("<td>%v</td>\n", d.TotalPVD)
			reponse += "</tr>\n"
		}
		_, err = rndr.Execute(c, map[string]interface{}{
			"tbody": reponse,
		})
		return err
	}
}

func ReadFile(fname string) (text string) {
	content, err := ioutil.ReadFile(fname)
	if err != nil {
		text = ""
		fmt.Printf("can't read file: %v\n", fname)
		log.Fatalln(err)
		return
	}

	// Convert []byte to string and print to screen
	text = string(content)
	return
}
